core/server/data/migrations/versions/4.42/2022-03-28-15-25-backfill-mrr-adjustments-for-offers.js

```diff
@@ -0,0 +1,170 @@
+                id: event.id,
+                member_id: event.member_id,
+                from_plan: event.from_plan,
+                to_plan: event.to_plan,
+                currency: event.currency,
+                source: event.source,
+                created_at: event.created_at,
+                mrr_delta: event.mrr_delta + mrrAdjustment
+            });
+        }
+
+        function getFirstEvent(events, redemption) {
+            const intervals = events.map(event => Interval.fromDateTimes(event.datetime, redemption.datetime));
+
+            // Invalid intervals would be if the end date was before the start date - e.g. offer redeemed before an MRR event
+            const validIntervals = intervals.filter(interval => interval.isValid);
+
+            if (validIntervals.length === 1) {
+                return events[intervals.indexOf(validIntervals[0])];
+            }
+
+            // At this point we have multiple possible first events, so we should check which is most likely to be correct based on timestamps
+            // Picking the closest one to when the redemption occured
+
+            // This is a butters way of getting the smallest probs better to do a sort or a loop or smth
+            const smallestIntervalLength = Math.min(...validIntervals.map(interval => interval.length()));
+            return events[intervals.indexOf(intervals.find(interval => interval.length() === smallestIntervalLength))];
+        }
+
+        offerRedemptions.forEach((redemption) => {
+            redemption.datetime = DateTime.fromISO(redemption.created_at);
+
+            const possibleEvents = mrrEventsByMemberId[redemption.member_id];
+
+            const firstEvents = possibleEvents.filter(event => event.from_plan === null && event.used !== true);
+
+            // If there is a single first event, then we know for sure that we need to
+            // 1. decrease MRR for it
+            // 2. increase MRR for a second event if it exists
+            if (firstEvents.length === 1) {
+                const firstEvent = firstEvents[0];
+                const mrrAdjustment = getMRRAdjustment(firstEvent, redemption);
+
+                updateEvent(firstEvent, -mrrAdjustment);
+
+                const secondEvent = possibleEvents.find(event => event.from_plan === firstEvent.to_plan);
+
+                if (secondEvent) {
+                    updateEvent(secondEvent, +mrrAdjustment);
+                }
+
+                return;
+            }
+
+            const firstEvent = getFirstEvent(firstEvents, redemption);
+            const mrrAdjustment = getMRRAdjustment(firstEvent, redemption);
+
+            updateEvent(firstEvent, -mrrAdjustment);
+
+            const mustHaveSecondEvent = redemption.subscription_status === 'canceled' || firstEvent.to_plan !== redemption.subscription_price;
+
+            const likelyDoesNotHaveSecondEvent = firstEvent.to_plan === redemption.subscription_price;
+
+            if (likelyDoesNotHaveSecondEvent) {
```

Comment:
Not sure, but might be worth to only skip if `mustHaveSecondEvent` is also false? Maybe that never happens, but I'm not sure.
```suggestion
            if (likelyDoesNotHaveSecondEvent && !mustHaveSecondEvent) {
```